stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -0,0 +1,376 @@
+    if (auto *ClassMetadata = *getIndirectClass())
+      return swift_getObjCClassMetadata(ClassMetadata);
+    else
+      return nullptr;
+  case TypeMetadataRecordKind::UniqueDirectClass:
+    if (auto *ClassMetadata = getDirectClass())
+      return swift_getObjCClassMetadata(ClassMetadata);
+    else
+      return nullptr;
+  default:
+    return nullptr;
+  }
+}
+
+// returns the type metadata for the type named by typeName
+static const Metadata *
+_searchTypeMetadataRecords(const TypeMetadataState &T,
+                           const llvm::StringRef typeName) {
+  unsigned sectionIdx = 0;
+  unsigned endSectionIdx = T.SectionsToScan.size();
+
+  for (; sectionIdx < endSectionIdx; ++sectionIdx) {
+    auto &section = T.SectionsToScan[sectionIdx];
+    for (const auto &record : section) {
+      if (auto metadata = record.getCanonicalTypeMetadata()) {
+        auto ntd = metadata->getNominalTypeDescriptor();
+
+        assert(ntd != nullptr);
+
+        if (typeName == ntd->Name) {
+          return metadata;
+        }
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+static const Metadata *
+_typeByMangledName(const llvm::StringRef typeName) {
+  const Metadata *foundMetadata = nullptr;
+  auto &T = TypeMetadataRecords.get();
+  size_t hash = llvm::HashString(typeName);
+
+  ConcurrentList<TypeMetadataCacheEntry> &Bucket = T.Cache.findOrAllocateNode(hash);
+
+  // Check name to type metadata cache
+  for (auto &Entry : Bucket) {
+    if (Entry.matches(typeName))
+      return Entry.getMetadata();
+  }
+
+  // Check type metadata records
+  pthread_mutex_lock(&T.SectionsToScanLock);
+  foundMetadata = _searchTypeMetadataRecords(T, typeName);
+  pthread_mutex_unlock(&T.SectionsToScanLock);
+
+  // Check protocol conformances table
+  if (foundMetadata == nullptr)
+    foundMetadata = _searchConformancesByMangledTypeName(typeName);
+
+  if (foundMetadata != nullptr)
+    Bucket.push_front(TypeMetadataCacheEntry(typeName, foundMetadata));
+
+#if SWIFT_OBJC_INTEROP
+  // Check for ObjC class
+  // FIXME does this have any value? any ObjC class with a Swift name
+  // should already be registered as a Swift type.
+  if (foundMetadata == nullptr) {
+    std::string prefixedName("_Tt" + typeName.str());
+    foundMetadata = reinterpret_cast<ClassMetadata *>
+      (objc_lookUpClass(prefixedName.c_str()));
+  }
+#endif
+
+  return foundMetadata;
+}
+
+/// Return the type metadata for a given mangled name, used in the
+/// implementation of _typeByName(). The human readable name returned
+/// by swift_getTypeName() is non-unique, so we used mangled names
+/// internally.
+extern "C"
+const Metadata *
+swift_getTypeByMangledName(const char *typeName, size_t typeNameLength) {
+  llvm::StringRef name(typeName, typeNameLength);
+  return _typeByMangledName(name);
+}
+
+/// Register a single type metadata instance in the name lookup cache, used
+/// to register instantiated generic types.
+void
+swift::_registerTypeMetadata(const Metadata *metadata) {
```

Comment:
Sure. If I can't figure out how to get the NTD and witness tables from a pattern I'll remove generic support. (After all the public API is limited to top level classes only.) But it'd be nice to get the plumbing complete from the start.

Sent from my iPhone

> On 12 Jan 2016, at 09:30, Joe Groff <notifications@github.com> wrote:
> 
> In stdlib/public/runtime/MetadataLookup.cpp:
> 
> > +
> > +/// Return the type metadata for a given mangled name, used in the
> > +/// implementation of _typeByName(). The human readable name returned
> > +/// by swift_getTypeName() is non-unique, so we used mangled names
> > +/// internally.
> > +extern "C"
> > +const Metadata *
> > +swift_getTypeByMangledName(const char *typeName, size_t typeNameLength) {
> > +  llvm::StringRef name(typeName, typeNameLength);
> > +  return _typeByMangledName(name);
> > +}
> > +
> > +/// Register a single type metadata instance in the name lookup cache, used
> > +/// to register instantiated generic types.
> > +void
> > +swift::_registerTypeMetadata(const Metadata *metadata) {
> I still say it's fine not to support generics for now.
> 
> —
> Reply to this email directly or view it on GitHub.
>